cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LocalizationUtils.java

```diff
@@ -0,0 +1,152 @@
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime;
+
+import co.cask.cdap.internal.app.runtime.distributed.LocalizeResource;
+import com.google.common.io.Files;
+import com.google.common.io.InputSupplier;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Enumeration;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Utilities for file localization.
+ */
+public final class LocalizationUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalizationUtils.class);
+
+  /**
+   * Localizes the specified {@link LocalizeResource} in the specified {@link File targetDir} with the specified
+   * file name and returns the {@link File} pointing to the localized file.
+   *
+   * @param fileName the name to localize the file with
+   * @param resource the {@link LocalizeResource} to localize
+   * @param targetDir the directory to localize the resource in
+   * @return the {@link File} pointing to the localized file.
+   */
+  public static File localizeResource(String fileName, LocalizeResource resource, File targetDir) throws IOException {
+    File tempFile = new File(targetDir, fileName);
+    File fileToLocalize = new File(resource.getURI().getPath());
+    if (resource.isArchive()) {
+      LOG.debug("Uncompress file {} to {}", fileToLocalize, tempFile);
+      uncompress(fileToLocalize, tempFile);
+    } else {
+      LOG.debug("Copy file from {} to {}", fileToLocalize, tempFile);
+      Files.copy(fileToLocalize, tempFile);
+    }
+    return tempFile;
+  }
+
+  private static void uncompress(File archive, File targetDir) throws IOException {
+    if (!targetDir.exists()) {
+      //noinspection ResultOfMethodCallIgnored
+      targetDir.mkdir();
+    }
+    String extension = Files.getFileExtension(archive.getPath());
+    switch (extension) {
+      case "zip":
+        unzip(archive, targetDir);
+        break;
+      case "jar":
+        unzip(archive, targetDir);
+        break;
+      case "gz":
+        // gz is not recommended for archiving multiple files together. So we only support .tar.gz
+        untargz(archive, targetDir);
+        break;
+      case "tar":
+        untar(archive, targetDir);
+        break;
+      default:
+        throw new IllegalArgumentException(String.format("Unsupported compression type %s. Only 'zip', 'jar', " +
+                                                           "'gz', 'gzip', 'tar', 'tar.gz' are supported.",
+                                                         extension));
+    }
+  }
+
+  private static void unzip(File zipFile, File targetDir) throws IOException {
```

Comment:
We have a method in `BundleJarUtil.unpackProgramJar` that does this. Also, better rename that method to `unJar` or `unZip`.